### DISCUSSION TITLE
How to access scheduled jobs meeting given criteria?

### DISCUSSION THREAD

**Discussion by amo13** [bensheldon/good_job#1086] [General]
Upon user registration, I schedule jobs in the future for a few purposes like reminding the users to confirm their email addresses. Now if a user cancels their account leaving future jobs tied to that account due, I'd like to cancel them too.
I'd like to use an `after_destroy` callback on the user model to find and cancel these scheduled jobs. 

How do I find these scheduled jobs using `GoodJob` though?

The only way I see is using `GoodJob::Job.all` and filtering using a `select{...}` statement like
```
GoodJob::Job.all.select{|job| job.serialized_params["arguments"][0]["user_id"] == user.id}
```

I am not sure though how efficient that is once there are a lot of jobs in the database after a while in production. Is there any more elegant way to achieve what I am trying to do?

Any idea is greatly appreciated.

**Comment by bensheldon** [comment]
Hi! Thanks for asking. What you're describing is a very common need.

I want to discourage you from using scheduled jobs in this way, for exactly the reasons you describe:

- it's inefficient in the database 
- it presents a problem of Job cancellation

I've personally solved this problem by doing this:

- If there is just a single or very small/countable number of email follow-ups, I'll add a `follow_up_message_sent_at` column to the Users table. I'll then create a Job that will run `User.where(follow_up_message_sent_at: nil).where("created_at < ?", 2.days.ago).each { |user|  UserEmails.send_follow_up(user).deliver_now }`, and schedule that job to run every day (or even every hour for some messages) using GoodJob Cron.
- If there is many more, or an arbitrary number of messages, I'll create a separate UserMessages table that tracks when these messages are sent and then do the same pattern of running a series of queries to determine who needs to determine what messages _right now_, rather than trying to schedule them in the future.

---

But, if you really did want to do future scheduled jobs, some other options are:

- Just allow the job to run at the scheduled time, and just doublecheck that the user hasn't been deleted before trying to send the email (either pass the user_id, or `rescue_from(ActiveJob::DeserializationError)`)
- You can query the job arguments, you'll just need to use Postgres JSON operators (albeit they will be slow unless you add custom indexes)

**Reply by amo13** [reply]
Thank you so much for your clear and concise answer! It is of great help to me.
That was all I needed.
